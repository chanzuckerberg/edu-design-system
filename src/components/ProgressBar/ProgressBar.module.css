/*------------------------------------*\
    # PROGRESS BAR
\*------------------------------------*/

/**
 * Progress Bar
 */
.progress-bar {
  display: flex;
  gap: var(--eds-size-half);

  border: var(--eds-theme-color-border-neutral-subtle) solid
    var(--eds-border-width-sm);
  border-radius: var(--eds-border-radius-full);

  /* Prevents child segments from clipping outside the borders */
  overflow: hidden;

  height: 0.875rem;
  width: 100%;
  background-color: var(--eds-theme-color-background-neutral-medium);
}

.progress-bar__segment {
  background-color: var(--eds-theme-color-background-brand-primary-strong);
}
