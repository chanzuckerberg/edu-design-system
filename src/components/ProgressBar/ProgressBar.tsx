import React from 'react';
import { useUID } from 'react-uid';
import styles from './ProgressBar.module.css';
import Text from '../Text';

export type Props = {
  /**
   * CSS class names that can be appended to the component.
   */
  className?: string;
  /**
   * Overrides the default "currentValue / maxValue" in cases where a different caption may be desired.
   */
  caption?: string;
  /**
   * The current amount of progress that has been made.
   */
  currentValue: number;
  /**
   * Label associated with naming the progress bar.
   */
  label: string;
  /**
   * Max possible value to be represented by the progress bar.
   */
  maxValue: number;
  /**
   * The amount of segments that make up the progress made.
   * Defaults to 1 and recommended max number of segments is 10.
   */
  totalSegments?: number;
};

/**
 * BETA: This component is still a work in progress and is subject to change.
 *
 * `import {ProgressBar} from "@chanzuckerberg/eds";`
 *
 * Show the progression of a system operation: downloading, uploading, processing, etc., in a visual way
 */
export const ProgressBar = ({
  caption,
  className,
  label,
  currentValue,
  maxValue,
  totalSegments = 1,
  ...other
}: Props) => {
  // Creates list of segment divs depending on the number of totalSegments.
  const segments = Array.from({ length: totalSegments }, (_, index) => (
    <div
      className={styles['progress-bar__segment']}
      key={'progress-bar__segment-' + index}
    />
  ));

  // Creates a div to represent progress made with width currentProgress over maxProgress as a percent.
  // It is the parent to the segments.
  const segmentsWrapper = (
    <div
      className={styles['progress-bar__segments-wrapper']}
      style={{ width: (currentValue / maxValue) * 100 + '%' }}
    >
      {segments}
    </div>
  );

  const progressBarCaption = caption || currentValue + '/' + maxValue;

  const labelId = useUID();
  const captionId = useUID();

  return (
    <div className={className} {...other}>
      <div className={styles['progress-bar__label-wrapper']}>
        <label className={styles['progress-bar__label']} id={labelId}>
          {label}
        </label>
        <Text as="span" id={captionId} size="caption">
          {progressBarCaption}
        </Text>
      </div>
      <div
        aria-describedby={captionId}
        aria-labelledby={labelId}
        aria-valuemax={maxValue}
        aria-valuenow={currentValue}
        className={styles['progress-bar']}
        // INFO: All descendants of role="progressbar" are presentational
        role="progressbar"
      >
        {segmentsWrapper}
      </div>
    </div>
  );
};
