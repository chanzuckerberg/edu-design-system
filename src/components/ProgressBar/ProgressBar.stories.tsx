import { BADGE } from '@geometricpanda/storybook-addon-badges';
import type { StoryObj, Meta } from '@storybook/react';
import React from 'react';
import { ProgressBar } from './ProgressBar';
import Button from '../Button';
import ButtonGroup from '../ButtonGroup';

export default {
  title: 'Molecules/Progress/ProgressBar',
  component: ProgressBar,
  args: {
    label: 'Progress Bar Label',
    max: 100,
    segmentCount: 2,
    segmentValue: 10,
  },
  parameters: {
    badges: [BADGE.BETA],
  },
  decorators: [
    (Story) => (
      <div style={{ backgroundColor: 'white', padding: '0.5rem' }}>
        <Story />
      </div>
    ),
  ],
} as Meta<Args>;

type Args = React.ComponentProps<typeof ProgressBar>;

export const Default: StoryObj<Args> = {};

export const WithCustomCaption: StoryObj<Args> = {
  args: {
    caption: '20 %',
  },
};

export const Empty: StoryObj<Args> = {
  args: {
    segmentCount: 0,
  },
};

export const Complete: StoryObj<Args> = {
  render: () => (
    <div>
      <ProgressBar
        label="Progress Bar Label"
        max={100}
        segmentCount={10}
        segmentValue={10}
      />
      <ProgressBar
        label="Progress Bar Label"
        max={100}
        segmentCount={1}
        segmentValue={100}
      />
    </div>
  ),
};

const InteractiveProgressBar = () => {
  const [segmentValue, setSegmentValue] = React.useState(1);
  const [max, setMax] = React.useState(10);
  const [count, setCount] = React.useState(0);
  const increase = () => {
    if (count < 10) {
      setCount(count + 1);
    }
  };
  const decrease = () => {
    if (count > 0) {
      setCount(count - 1);
    }
  };
  return (
    <div>
      <ProgressBar
        label="Interactive progress bar"
        max={max}
        segmentCount={count}
        segmentValue={segmentValue}
      />
      <br />
      <label htmlFor="segment-value-input">Segment value: </label>
      <input
        id="segment-value-input"
        onChange={(e) => {
          setCount(0);
          setSegmentValue(Number(e?.target?.value));
        }}
        value={segmentValue}
      ></input>
      <br />
      <label htmlFor="max-input">Max value: </label>
      <input
        id="max-input"
        onChange={(e) => {
          setCount(0);
          setMax(Number(e?.target?.value));
        }}
        value={max}
      ></input>
      <br />
      <ButtonGroup>
        {count > 0 && (
          <Button onClick={decrease} variant="primary">
            Decrease Segment Count
          </Button>
        )}
        {count < 10 && (
          <Button onClick={increase} variant="secondary">
            Increase Segment Count
          </Button>
        )}
      </ButtonGroup>
    </div>
  );
};

export const Interactive: StoryObj<Args> = {
  render: () => <InteractiveProgressBar />,
  parameters: {
    /**
     * For interactive purposes only, low value in snapping or checking for visual regression since they should be covered in the other stories.
     */
    chromatic: {
      disableSnapshot: true,
    },
    snapshot: {
      skip: true,
    },
  },
};
