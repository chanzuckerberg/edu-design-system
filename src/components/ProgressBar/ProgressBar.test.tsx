import { generateSnapshots } from '@chanzuckerberg/story-utils';
import { render } from '@testing-library/react';
import React from 'react';
import { ProgressBar } from './ProgressBar';
import * as stories from './ProgressBar.stories';
import consoleWarnMockHelper from '../../../jest/helpers/consoleWarnMock';

describe('<ProgressBar />', () => {
  const warnMock = consoleWarnMockHelper();

  generateSnapshots(stories);
  it('should warn if segment count is larger than 10', () => {
    render(
      <ProgressBar
        label="Lorem Ipsum"
        max={100}
        segmentCount={11}
        segmentValue={1}
      />,
    );
    expect(warnMock).toHaveBeenCalledWith(
      'Progress bar segment count should be capped at 10',
    );
  });
});
