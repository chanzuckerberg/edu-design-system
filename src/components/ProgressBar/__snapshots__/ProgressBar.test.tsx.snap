// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar /> Complete story renders snapshot 1`] = `
<div
  style="background-color: white; padding: 0.5rem;"
>
  <div>
    <div>
      <div
        class="progress-bar__label-wrapper"
      >
        <label
          class="progress-bar__label"
          id="7"
        >
          Progress Bar Label
        </label>
        <span
          class="text text--caption"
          id="8"
        >
          100/100
        </span>
      </div>
      <div
        aria-describedby="8"
        aria-labelledby="7"
        aria-valuemax="100"
        aria-valuenow="100"
        class="progress-bar"
        role="progressbar"
      >
        <div
          class="progress-bar__segment"
          style="width: 10%;"
        />
        <div
          class="progress-bar__segment"
          style="width: 10%;"
        />
        <div
          class="progress-bar__segment"
          style="width: 10%;"
        />
        <div
          class="progress-bar__segment"
          style="width: 10%;"
        />
        <div
          class="progress-bar__segment"
          style="width: 10%;"
        />
        <div
          class="progress-bar__segment"
          style="width: 10%;"
        />
        <div
          class="progress-bar__segment"
          style="width: 10%;"
        />
        <div
          class="progress-bar__segment"
          style="width: 10%;"
        />
        <div
          class="progress-bar__segment"
          style="width: 10%;"
        />
        <div
          class="progress-bar__segment"
          style="width: 10%;"
        />
      </div>
    </div>
    <div>
      <div
        class="progress-bar__label-wrapper"
      >
        <label
          class="progress-bar__label"
          id="9"
        >
          Progress Bar Label
        </label>
        <span
          class="text text--caption"
          id="10"
        >
          100/100
        </span>
      </div>
      <div
        aria-describedby="10"
        aria-labelledby="9"
        aria-valuemax="100"
        aria-valuenow="100"
        class="progress-bar"
        role="progressbar"
      >
        <div
          class="progress-bar__segment"
          style="width: 100%;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ProgressBar /> Default story renders snapshot 1`] = `
<div
  style="background-color: white; padding: 0.5rem;"
>
  <div>
    <div
      class="progress-bar__label-wrapper"
    >
      <label
        class="progress-bar__label"
        id="1"
      >
        Progress Bar Label
      </label>
      <span
        class="text text--caption"
        id="2"
      >
        20/100
      </span>
    </div>
    <div
      aria-describedby="2"
      aria-labelledby="1"
      aria-valuemax="100"
      aria-valuenow="20"
      class="progress-bar"
      role="progressbar"
    >
      <div
        class="progress-bar__segment"
        style="width: 10%;"
      />
      <div
        class="progress-bar__segment"
        style="width: 10%;"
      />
    </div>
  </div>
</div>
`;

exports[`<ProgressBar /> Empty story renders snapshot 1`] = `
<div
  style="background-color: white; padding: 0.5rem;"
>
  <div>
    <div
      class="progress-bar__label-wrapper"
    >
      <label
        class="progress-bar__label"
        id="5"
      >
        Progress Bar Label
      </label>
      <span
        class="text text--caption"
        id="6"
      >
        0/100
      </span>
    </div>
    <div
      aria-describedby="6"
      aria-labelledby="5"
      aria-valuemax="100"
      aria-valuenow="0"
      class="progress-bar"
      role="progressbar"
    />
  </div>
</div>
`;

exports[`<ProgressBar /> WithUnit story renders snapshot 1`] = `
<div
  style="background-color: white; padding: 0.5rem;"
>
  <div>
    <div
      class="progress-bar__label-wrapper"
    >
      <label
        class="progress-bar__label"
        id="3"
      >
        Progress Bar Label
      </label>
      <span
        class="text text--caption"
        id="4"
      >
        20/100 %
      </span>
    </div>
    <div
      aria-describedby="4"
      aria-labelledby="3"
      aria-valuemax="100"
      aria-valuenow="20"
      aria-valuetext="%"
      class="progress-bar"
      role="progressbar"
    >
      <div
        class="progress-bar__segment"
        style="width: 10%;"
      />
      <div
        class="progress-bar__segment"
        style="width: 10%;"
      />
    </div>
  </div>
</div>
`;
